Basket feature

VeniNovakov opened the pull request.

VeniNovakov committed:
now adding and subtracting from the basket values as well as removing items from it

VeniNovakov committed:
fixed a bug in chrome

bug when writing to a cookie(object instead of list of objects as a string were written)

VeniNovakov committed:
removing amount when amount is 1 is disabled

VeniNovakov committed:
Added product page version with no functionality for now and some minor changes to Menu page and basket page

VeniNovakov committed:
products can be added from the product page

NikolaPavlovski002 commented on pizzeria-frontend/src/Menu.tsx:
```diff
@@ -1,100 +1,136 @@
+
+  const filteredMenu = selectedFilter
+    ? menu.filter((product) => product.category === selectedFilter)
+    : menu;
```
Move this to useEffect with dependancy prop selectedFilter, as this way it would update only on re-render. ( Could still work but unclear )

VeniNovakov replied:
done

NikolaPavlovski002 commented on pizzeria-frontend/src/Menu.tsx:
```diff
@@ -1,100 +1,136 @@
 };
 
-const Filter = () => {
+const Filter = ({ onFilterClick }: any) => {
```
Remove **any**, you can create an interface with {
 onFilterClick: (_: string) => void;
 }
 and so it knows its getting a function with props and output

VeniNovakov replied:
Done

NikolaPavlovski002 commented on pizzeria-frontend/src/Menu.tsx:
```diff
@@ -1,100 +1,136 @@
 };
 
-const FilterItem = (props: any) => {
+const FilterItem = ({ item, onFilterClick }: any) => {
```
Same here, remove any

VeniNovakov replied:
Done

NikolaPavlovski002 commented on pizzeria-frontend/src/Menu.tsx:
```diff
@@ -1,100 +1,136 @@
     const newBasketObj: BasketItem = { product: product, amount: 1 };
 
     if (basketItems.length === 0) {
+      console.log(basketItems);
```
might want to remove this, if its intentional just resolve this thread

VeniNovakov replied:
Done

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/ProductPage.tsx:
```diff
@@ -1,26 +1,208 @@
+  const removeAmount = () => {
+    if (amount > 1 && instanceOfProduct(item)) {
+      amount--;
+      setAmount(amount); 
```
Why are you doing 
`amount--;`
and then
`setAmount(amount); `
Cant you just do `setAmount(amount - 1)` ?

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/ProductPage.tsx:
```diff
@@ -1,26 +1,208 @@
+    }
+
+    if(amount === 1 && instanceOfProduct(item)){
+      console.log('dsadsad');
```
Add a TODO if this is not part of this PR's task

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/basketPage.tsx:
```diff
@@ -1,35 +1,207 @@
+
+    Cookies.set("basket", basketString);
+
+  }, [basketItems, setBasketCounter]);
```
Have you tested whether this calls itself permanently as i have a suspicion it might?

VeniNovakov replied:
Tested it now, it does not console log endlessly if that means something. I think it could have happened if I changed the basketItems array in this useEffect though

NikolaPavlovski002 replied:
Yup, if it doesnt consoleLog endlessly its good

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/basketPage.tsx:
```diff
@@ -1,35 +1,207 @@
+const ProductOrdered = (props: BasketProps) => {
+  const { basketItems, setBasketItems } = useBasketContent();
+  const { setBasketCounter } = useBasket();
+  const IsButtonDisabled = props.item.amount === 1 ? true:false
```
Would put it in a useEffect again,
useEffect(() => {
  setIsButtonDisabled(props.item.amount === 1 );
  }, [props.item.amount]);

VeniNovakov replied:
Done

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/basketPage.tsx:
```diff
@@ -1,35 +1,207 @@
+        </>
+      ) : (
+        <div></div>
+      )}
```
You could change the ? to &&
````
{basketItems.length && (
        <>
          <Total></Total>
          <button
            type="button"
            className="shadow-md hover:shadow-inner hover:bg-slate-100 rounded-md border-slate-400 border"
          >
            ORDER
          </button>
        </>
      )}

VeniNovakov replied:
Done

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/ProductPage.tsx:
```diff
@@ -1,26 +1,208 @@
+    //when clicked it will add it to the basket cookie
+    const tempProduct = JSON.parse(Cookies.get('tempProduct') || '');
+    Cookies.set('basket', JSON.stringify([...basketItems, tempProduct]));
+    setBasketItems(JSON.parse(Cookies.get('basket') || '[]'));
```
Optional,
As in every place you do setBasketItems, then you do Cookies.set('basket',
It would be a lot of refactoring so thats why im saying its optional, but it would make it clearer code. 
You could add in the Basket Context an use Effect or directly in the setBascetItems to add it to the cookies so you dont have to worry every time to add it

VeniNovakov replied:
Done. Chose to add the useEffect in the Basket Context as it sounds like the better option

NikolaPavlovski002 commented on pizzeria-frontend/src/pages/ProductPage.tsx:
```diff
@@ -1,26 +1,208 @@
+        Cookies.set('tempProduct', JSON.stringify({...ck, addOns: [...updatedAddOns]}));
+      }
+    }
+    }, [amount])
```
This is quite a big useEffect and hard to read, would split it in couple if functions and every if would call its function then return. That way here you have all the logic like
if(Cookies.get('tempProduct')){
  doCookieUpdatedAddOns()
  return;
 }
 And if someone needs to see what the this if does he just goes in the function and looks

VeniNovakov replied:
Done. In the new commit the useEffect is way smaller and contains much smaller part of the logic. Improved readability(could probably be better also). The logic of updating the amount of product/addOn is now handled by other functions

VeniNovakov committed:
fixes

total is now displayed properly, can add items with add-ons and their value is also calculated properly and the add-ons are displayed on hover, bug fixes when page loading

VeniNovakov committed:
addressed PR comments and refactored code

NikolaPavlovski002 commented on pizzeria-frontend/src/Menu.tsx:
```diff
@@ -1,56 +1,102 @@
+  const { basketCounter, setBasketCounter } = useBasket();
 
   useEffect(() => {
     Cookies.set("basket", JSON.stringify(basketItems));
```
Now that you have the useEffect, you can skip this here

NikolaPavlovski002 commented:
Looks good, you can merge
